Fix user scope rules

PiDelport opened the pull request:
Closes #348

longtomjr reviewed (changes requested):
Just a couple of minor things

PiDelport commented on src/rules.spec.ts:
```diff
@@ -0,0 +1,85 @@
+  ctx: Record<string, unknown>,
+): Promise<boolean | string | Error> {
+  // See upstream rule tests for this invocation pattern:
+  // https://github.com/maticzav/graphql-shield/blob/master/tests/logic.test.ts
```
Good catch, thanks!

PiDelport commented on src/rules.ts:
```diff
@@ -11,29 +11,29 @@ const isEventQueryOwner = rule({
+const _userHasScope: (scopeToken: string) => IRule = scopeToken =>
+  rule({ cache: 'contextual' })(
+    (_parent, _args, { req }: ContextWithUser, _info) => {
+      const scope = req?.user?.scope;
```
Cool. I wasn't sure what context guarantees we can actually rely on here without a runtime guard: I know that some GraphQL implementations may not supply the `user` for anonymous queries?

(I'm trying to find a reference about what's actually guaranteed in our case.)

longtomjr replied:
@PiDelport The user is guaranteed since we add it to the request during authentication. [The gqlAuthMiddleware](https://github.com/registreerocks/registree-core/blob/staging/src/auth/gql-auth.middleware.ts#L8) (graphql middleware in this case) runs and use passport to validate the token, and adds the user to the request context. You can see where it is configured to run [in the graphql config of the app.module.ts](https://github.com/registreerocks/registree-core/blob/5a8612329044b5621d1ae9f10243e4086c5fd382/src/app.module.ts#L71-L75)

Any anonymous queries will fail because of the middleware since we require authentication before any query proceeds.

PiDelport replied:
Yup, the user should normally be there: the checks here are more defensive, in case of misconfiguration or bugs.

Better error reporting would definitely be good, though, yeah: see the https://github.com/registreerocks/registree-core/pull/350#issuecomment-763644012 update.

PiDelport commented on src/rules.ts:
```diff
@@ -11,29 +11,29 @@ const isEventQueryOwner = rule({
+    (_parent, _args, { req }: ContextWithUser, _info) => {
+      const scope = req?.user?.scope;
+      return (
+        (scope !== undefined && _scopeContains(scope, scopeToken)) ||
```
Hmm, I wanted to keep the check narrowed down to `undefined` from the `?.` operators to avoid swallowing an unexpected `null` scope here, but yeah, it's messy. A reusable helper to reliably get the user from scope would make sense here, and simplify the code.

longtomjr replied:
If we are certain (and I am not sure on this) that scope is required for a valid user context, we might just want to error. I think we want the type-system to deal with this. If you remove the `.?` chaining, and the type is just `string` and not `string | null | undefined` (or just 2 of them) then we should not be doing a null check here.

We might want to change the `User` type however to match up with the standards if null/undefined is possible, but I would prefer if we create a task for that and action it separately from this PR.

PiDelport replied:
The https://github.com/registreerocks/registree-core/pull/350#issuecomment-763644012 update should deal with this better.

PiDelport commented:
Okay, I tried a couple of things, and reworked this a bit.

I introduced a new `userFromContext` helper for checked access to the user from context. I refactored the code to use it, and also introduced it in the `@CurrentUser` decorator, which does the same kind of check.

The code is somewhat defensive, but it should be worth it here: if there's any misconfiguration, bug, or upstream issue in the auth middleware, a helpful error message and traceback from here will probably be very helpful.

longtomjr commented on src/rules.ts:
```diff
@@ -1,56 +1,67 @@
+ *
+ * @throw {Error} If the context has no user.
+ */
+export function userFromContext(ctx: { req?: { user?: User } }): User {
```
Maybe rename this to `userFromGraphqlContext` or something similar. For helpers like this is makes sense to add it to a file in the src/common directory, or create a new file there if there is no relevant files available.

PiDelport replied:
How about `userFromResolverContext`? (GraphQL has enough different types of contexts in code that it's useful for me to try and make it obvious which one.)

longtomjr replied:
`userFromResolverContext` or `userFromGqlResolverContext` are both fine with me. I think I prefer the second one, but it is probably redundant to have `Gql` and `Resolver` (since resolver already indicates that we are likely working with graphql).

PiDelport replied:
Cool, I like `Gql`. 😊️

(The "resolver context" still helps disambiguate it from the other types of context in GraphQL, but it should be fine to delegate that to the doc comment rather than the name, yeah.)

PiDelport replied:
Alright, I moved the helper to `userFromGqlContext` under `common/graphql.helpers`: does that look good?

longtomjr reviewed (changes requested):
Just the one change and then we can get this merged.

PiDelport commented:
Thanks!
